src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs

```diff
@@ -21,16 +22,24 @@ internal sealed class MetadataAsSourceGeneratedFileInfo
 
         private readonly ParseOptions _parseOptions;
 
-        public MetadataAsSourceGeneratedFileInfo(string rootPath, Project sourceProject, INamedTypeSymbol topLevelNamedType)
+        public MetadataAsSourceGeneratedFileInfo(string rootPath, Project sourceProject, INamedTypeSymbol topLevelNamedType, bool allowDecompilation)
         {
             this.SourceProjectId = sourceProject.Id;
-            _parseOptions = sourceProject.ParseOptions;
             this.Workspace = sourceProject.Solution.Workspace;
-            this.LanguageName = sourceProject.Language;
+            if (allowDecompilation && sourceProject.Language != LanguageNames.CSharp)
```

Comment:
If you just move this after this.LanguageName =, then I think you can reduce this all back to _parseOptions = Workspace.Services.GetLanguageServices(this.LanguageName).GetRequriedService().ParseOptions;. Basically, you've already figured out once that if it's decompiled it's always C#; no reason to do it twice.